linter/lints/common.go

```diff
@@ -35,3 +37,50 @@ func GetExtWithOID(exts []pkix.Extension, oid asn1.ObjectIdentifier) *pkix.Exten
 	}
 	return nil
 }
+
+// ReadOptionalASN1BooleanWithTag attempts to read and advance incoming to
+// search for an optional DER-encoded ASN.1 element tagged with the given tag.
+// Unless outExpectedTagPresent or outPresent are nil, it stores whether the tag
+// was found in outExpectedTagPresent and an element with the tag was found in
+// outPresent, otherwise each boolean will take the default value. It reports
+// whether all reads were successful.
+func ReadOptionalASN1BooleanWithTag(incoming *cryptobyte.String, outExpectedTagPresent *bool, outPresent *bool, tag cryptobyte_asn1.Tag, defaultValue bool) bool {
+	if outExpectedTagPresent != nil {
+		*outExpectedTagPresent = defaultValue
+	}
+	if outPresent != nil {
+		*outPresent = defaultValue
+	}
+	if incoming.Empty() {
+		return false
+	}
+
+	// We need to check the boolean value to determine if the tag was found. The
+	// ReadOptionalASN1BooleanWithTag caller may not care about the value, but
+	// here internally we do.
+	tagPresent := false
+	var tagBytes cryptobyte.String
+
+	// ReadOptionalASN1 performs a peek and will not advance if the tag is
+	// missing, meaning that incoming will retain bytes.
+	ok := incoming.ReadOptionalASN1(&tagBytes, &tagPresent, tag)
+	if !ok && !tagPresent {
```

Comment:
This returns false only if the read failed, *and* the tag was not present. That doesn't make sense; if the read failed we don't really know if the tag was present.

The general pattern we should follow here is:

 - If the read failed, return false without touching `*out`
 - If the read succeeded, but the tag was not present, write `defaultValue` to `*out` and return true.
 - If the read succeeded, and the tag _was_ present, compare the value against `0xFF`.